# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/opengl/TextureClientOGL.cpp
# Commit: 290917deea87
# Full Hash: 290917deea879f965f6f1ad441df945c41c01c14
# Author: Jamie Nicol <jnicol@mozilla.com>
# Date: 2023-12-13 21:42:49
# Regressor Bug: 1866020
# File Overlap Count: 1
# Description:
#   Bug 1866020 - Override buggy colorspace conversion on Pixel devices. r=gw,padenot,geckoview-reviewers,owlish
#   
#   Pixel 6, 7, and 8 devices running Android 14 are affected by a bug
#   where video frames with SMPTE 432 color primaries are rendered
#   incorrectly when sampled from an external texture in GLES. To work
# ==============================================================================

diff -r 12e266e296c1 -r 290917deea87 gfx/layers/opengl/TextureClientOGL.cpp
--- a/gfx/layers/opengl/TextureClientOGL.cpp	Wed Dec 13 08:29:00 2023 +0000
+++ b/gfx/layers/opengl/TextureClientOGL.cpp	Wed Dec 13 09:06:29 2023 +0000
@@ -38,7 +38,7 @@
 
 already_AddRefed<TextureClient> AndroidSurfaceTextureData::CreateTextureClient(
     AndroidSurfaceTextureHandle aHandle, gfx::IntSize aSize, bool aContinuous,
-    gl::OriginPos aOriginPos, bool aHasAlpha,
+    gl::OriginPos aOriginPos, bool aHasAlpha, bool aForceBT709ColorSpace,
     Maybe<gfx::Matrix4x4> aTransformOverride, LayersIPCChannel* aAllocator,
     TextureFlags aFlags) {
   if (aOriginPos == gl::OriginPos::BottomLeft) {
@@ -47,17 +47,19 @@
 
   return TextureClient::CreateWithData(
       new AndroidSurfaceTextureData(aHandle, aSize, aContinuous, aHasAlpha,
-                                    aTransformOverride),
+                                    aForceBT709ColorSpace, aTransformOverride),
       aFlags, aAllocator);
 }
 
 AndroidSurfaceTextureData::AndroidSurfaceTextureData(
     AndroidSurfaceTextureHandle aHandle, gfx::IntSize aSize, bool aContinuous,
-    bool aHasAlpha, Maybe<gfx::Matrix4x4> aTransformOverride)
+    bool aHasAlpha, bool aForceBT709ColorSpace,
+    Maybe<gfx::Matrix4x4> aTransformOverride)
     : mHandle(aHandle),
       mSize(aSize),
       mContinuous(aContinuous),
       mHasAlpha(aHasAlpha),
+      mForceBT709ColorSpace(aForceBT709ColorSpace),
       mTransformOverride(aTransformOverride) {
   MOZ_ASSERT(mHandle);
 }
@@ -76,7 +78,7 @@
   aOutDescriptor = SurfaceTextureDescriptor(
       mHandle, mSize,
       mHasAlpha ? gfx::SurfaceFormat::R8G8B8A8 : gfx::SurfaceFormat::R8G8B8X8,
-      mContinuous, mTransformOverride);
+      mContinuous, mForceBT709ColorSpace, mTransformOverride);
   return true;
 }
 
@@ -159,6 +161,7 @@
     SurfaceDescriptor& aOutDescriptor) {
   aOutDescriptor = SurfaceTextureDescriptor(
       mSurface->GetHandle(), mSize, mFormat, false /* not continuous */,
+      false /* do not override colorspace */,
       Some(gfx::Matrix4x4()) /* always use identity transform */);
   return true;
 }